api/src/main/java/org/pmiops/workbench/mail/MailServiceImpl.java

```diff
@@ -1,16 +1,120 @@
+import org.pmiops.workbench.mandrill.model.MandrillMessageStatus;
+import org.pmiops.workbench.mandrill.model.RecipientAddress;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
+import java.util.Collections;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.inject.Provider;
 import javax.mail.Message;
 import javax.mail.MessagingException;
 import javax.mail.Transport;
 
 @Service
 public class MailServiceImpl implements MailService {
 
+    private final Provider<MandrillApi> mandrillApiProvider;
+    private final Provider<CloudStorageService> cloudStorageServiceProvider;
+    private Provider<WorkbenchConfig> workbenchConfigProvider;
+    private static final Logger log = Logger.getLogger(MailServiceImpl.class.getName());
+    enum Status {REJECTED, API_ERROR, SUCCESSFUL}
+
+    @Autowired
+    public MailServiceImpl(Provider<MandrillApi> mandrillApiProvider,
+                           Provider<CloudStorageService> cloudStorageServiceProvider,
+                           Provider<WorkbenchConfig> workbenchConfigProvider) {
+        this.mandrillApiProvider = mandrillApiProvider;
+        this.cloudStorageServiceProvider = cloudStorageServiceProvider;
+        this.workbenchConfigProvider = workbenchConfigProvider;
+    }
+
     public void send(Message msg) throws MessagingException {
         Transport.send(msg);
     }
 
+    public void sendWelcomeEmail(String contactEmail, String password, User user) throws MessagingException {
+        String apiKey = cloudStorageServiceProvider.get().readMandrillApiKey();
+        int retries = workbenchConfigProvider.get().mandrill.sendRetries;
+        MandrillApiKeyAndMessage keyAndMessage = new MandrillApiKeyAndMessage();
+        keyAndMessage.setKey(apiKey);
+        MandrillMessage msg = buildWelcomeMessage(contactEmail, password, user);
+        keyAndMessage.setMessage(msg);
+        retry: do {
+            retries--;
+            log.log(Level.INFO, "retries: " + retries);
+            ImmutablePair attempt = trySend(keyAndMessage);
+            Status status = Status.valueOf(attempt.getLeft().toString());
+            switch (status) {
+                case API_ERROR:
+                    log.log(Level.WARNING, String.format(
+                      "ApiException: Welcome Email to '%s' for user '%s' not sent: %s", contactEmail, user.getName(), attempt.getRight().toString()));
+                    if (retries == 0) {
+                        log.log(Level.SEVERE, String.format(
+                          "ApiException: On Last Attempt! Welcome Email to '%s' for user '%s' not sent: %s", contactEmail, user.getName(), attempt.getRight().toString()));
+                        throw new MessagingException("Sending email failed: " + attempt.getRight().toString());
+                    }
+                    break;
+
+                case REJECTED:
```

Comment:
So "REJECTED" here means that a retry is bound to fail?

Reply:
Will defer to @UrsaStutsman on this, but I think it is true that rejected will always fail. This means that mandrill didn't like what we asked it to send, or didn't like our api key, or our license is not valid ... etc.

Reply:
Yes - "REJECTED" means we communicated with Mandrill, but Mandrill rejected it. This can be for any number of reasons, this is why I'm logging the reject_reason from mandrill. Common reasons, email gets a "hard_bounce", ie does not exist, something is wrong with our api key, etc...